File: codex_poting/java/src/main/java/com/example/demo/config/LoginRateLimitService.java
package com.example.demo.config;

import com.example.demo.exception.LoginRateLimitExceededException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.time.Clock;
import java.time.LocalDate;
import java.util.concurrent.ConcurrentHashMap;

@Component
public class LoginRateLimitService {
    private final LoginRateLimitProperties properties;
    private final Clock clock;
    private final ConcurrentHashMap<String, LoginAttemptCounter> counters = new ConcurrentHashMap<>();

    @Autowired
    public LoginRateLimitService(LoginRateLimitProperties properties) {
        this(properties, Clock.system(properties.getZoneId()));
    }

    LoginRateLimitService(LoginRateLimitProperties properties, Clock clock) {
        this.properties = properties;
        this.clock = clock;
    }

    public void consume(String username) {
        if (!properties.isEnabled()) {
            return;
        }

        final var today = LocalDate.now(clock.withZone(properties.getZoneId()));
        counters.compute(username, (key, counter) -> nextCounter(counter, today));
    }

    public void resetAll() {
        counters.clear();
    }

    private LoginAttemptCounter nextCounter(LoginAttemptCounter counter, LocalDate today) {
        if (counter == null || !counter.date().equals(today)) {
            return new LoginAttemptCounter(today, 1);
        }
        if (counter.count() >= properties.getDailyLimit()) {
            throw new LoginRateLimitExceededException();
        }
        return new LoginAttemptCounter(today, counter.count() + 1);
    }

    private record LoginAttemptCounter(LocalDate date, int count) {
    }
}

File: codex_poting/kotlin/src/main/kotlin/com/example/demo/config/LoginRateLimitService.kt
package com.example.demo.config

import com.example.demo.exception.LoginRateLimitExceededException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.time.Clock
import java.time.LocalDate
import java.util.concurrent.ConcurrentHashMap

@Component
class LoginRateLimitService internal constructor(
    private val properties: LoginRateLimitProperties,
    private val clock: Clock
) {
    private val counters: ConcurrentHashMap<String?, LoginAttemptCounter?> =
        ConcurrentHashMap<String?, LoginAttemptCounter?>()

    @Autowired
    constructor(properties: LoginRateLimitProperties) : this(properties, Clock.system(properties.zoneId))

    fun consume(username: String?) {
        if (!properties.isEnabled()) {
            return
        }

        val today = LocalDate.now(clock.withZone(properties.zoneId))
        counters.compute(username) { key: String?, counter: LoginAttemptCounter? -> nextCounter(counter, today) }
    }

    fun resetAll() {
        counters.clear()
    }

    private fun nextCounter(counter: LoginAttemptCounter?, today: LocalDate?): LoginAttemptCounter {
        if (counter == null || counter.date != today) {
            return LoginAttemptCounter(today, 1)
        }
        if (counter.count >= properties.dailyLimit) {
            throw LoginRateLimitExceededException()
        }
        return LoginAttemptCounter(today, counter.count + 1)
    }

    private data class LoginAttemptCounter(val date: LocalDate?, val count: Int)
}
